From eca11245b9f0c573522105a780edd063d9eaf667 Mon Sep 17 00:00:00 2001
From: Alex Williamson <alex.williamson@redhat.com>
Date: Mon, 16 Jun 2014 19:48:43 -0400
Subject: [kernel] tick: Clear broadcast pending bit when switching to oneshot

Message-id: <20140616194811.3200.41262.stgit@bling.home>
Patchwork-id: 84113
O-Subject: [RHEL7.1/7.0.z kernel PATCH] tick: Clear broadcast pending bit when switching to oneshot
Bugzilla: 1109996
RH-Acked-by: Myron Stowe <mstowe@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Stanislaw Gruszka <sgruszka@redhat.com>

From: Thomas Gleixner <tglx@linutronix.de>

Bugzilla: 1109996
Brew: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=7584787
Upstream commit: dd5fd9b91a77b4c9c28b7ef9c181b1a875820d0a

AMD systems which use the C1E workaround in the amd_e400_idle routine
trigger the WARN_ON_ONCE in the broadcast code when onlining a CPU.

The reason is that the idle routine of those AMD systems switches the
cpu into forced broadcast mode early on before the newly brought up
CPU can switch over to high resolution / NOHZ mode. The timer related
CPU1 bringup looks like this:

  clockevent_register_device(local_apic);
  tick_setup(local_apic);
  ...
  idle()
 tick_broadcast_on_off(FORCE);
 tick_broadcast_oneshot_control(ENTER)
   cpumask_set(cpu, broadcast_oneshot_mask);
 halt();

Now the broadcast interrupt on CPU0 sets CPU1 in the
broadcast_pending_mask and wakes CPU1. So CPU1 continues:

 local_apic_timer_interrupt()
    tick_handle_periodic();
    softirq()
      tick_init_highres();
        cpumask_clr(cpu, broadcast_oneshot_mask);

 tick_broadcast_oneshot_control(ENTER)
    WARN_ON(cpumask_test(cpu, broadcast_pending_mask);

So while we remove CPU1 from the broadcast_oneshot_mask when we switch
over to highres mode, we do not clear the pending bit, which then
triggers the warning when we go back to idle.

The reason why this is only visible on C1E affected AMD systems is
that the other machines enter the deep sleep states via
acpi_idle/intel_idle and exit the broadcast mode before executing the
remote triggered local_apic_timer_interrupt. So the pending bit is
already cleared when the switch over to highres mode is clearing the
oneshot mask.

The solution is simple: Clear the pending bit together with the mask
bit when we switch over to highres mode.

Stanislaw came up independently with the same patch by enforcing the
C1E workaround and debugging the fallout. I picked mine, because mine
has a changelog :)

Reported-by: poma <pomidorabelisima@gmail.com>
Debugged-by: Stanislaw Gruszka <sgruszka@redhat.com>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Cc: Olaf Hering <olaf@aepfle.de>
Cc: Dave Jones <davej@redhat.com>
Cc: Justin M. Forbes <jforbes@redhat.com>
Cc: Josh Boyer <jwboyer@redhat.com>
Link: http://lkml.kernel.org/r/alpine.DEB.2.02.1402111434180.21991@ionos.tec.linutronix.de
Cc: stable@vger.kernel.org # 3.10+
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/time/tick-broadcast.c b/kernel/time/tick-broadcast.c
index 6d3f916..19ee339 100644
--- a/kernel/time/tick-broadcast.c
+++ b/kernel/time/tick-broadcast.c
@@ -752,6 +752,7 @@ out:
 static void tick_broadcast_clear_oneshot(int cpu)
 {
  cpumask_clear_cpu(cpu, tick_broadcast_oneshot_mask);
+ cpumask_clear_cpu(cpu, tick_broadcast_pending_mask);
 }
 
 static void tick_broadcast_init_next_event(struct cpumask *mask,
-- 
1.7.1